### PR TITLE
feat: add IDA handles and type annotations

### PR DESCRIPTION
**Pull request by mr-tz**
and various other fixes

test in IDA pass, outstanding test fails are due to freeze, JSON, and dotnet issues

```
OK   mimikatz-function=0x40E5C2-basic block-7
OK   mimikatz-function=0x4702FD-characteristic(calls from)-0
OK   mimikatz-function=0x40E5C2-characteristic(calls from)-3
OK   mimikatz-function=0x4556E5-characteristic(calls to)-0
OK   mimikatz-function=0x40B1F1-characteristic(calls to)-3
SKIP 3b13b...-function=0x10006860-characteristic(nzxor)-True
SKIP 64d9f-function=0x10001510,bb=0x100015B0-offset(0x4000)-True
SKIP 7351f.elf-file-os(linux)-True
SKIP 7351f.elf-file-os(windows)-False
SKIP 7351f.elf-file-format(elf)-True
SKIP 7351f.elf-file-format(pe)-False
SKIP 7351f.elf-file-arch(i386)-False
SKIP 7351f.elf-file-arch(amd64)-True
SKIP 7351f.elf-function=0x408753-string(/dev/null)-True
SKIP 7351f.elf-function=0x408753,bb=0x408781-api(open)-True
SKIP 773290...-function=0x140001140-string(%s:\\OfficePackagesForWDAG)-True
SKIP 79abd...-function=0x10002385,bb=0x10002385-characteristic(call $+5)-True
SKIP 946a9...-function=0x10001510,bb=0x100015c0-characteristic(call $+5)-True
SKIP a1982...-function=0x4014D0-characteristic(cross section flow)-True
SKIP al-khaser x64-function=0x14004B4F0-api(__vcrt_GetModuleHandle)-True
SKIP c91887...-function=0x40156F-api(CloseClipboard)-True
SKIP c91887...-function=0x401A77-api(kernel32.CreatePipe)-True
SKIP c91887...-function=0x401A77-api(kernel32.SetHandleInformation)-True
SKIP c91887...-function=0x401A77-api(kernel32.CloseHandle)-True
SKIP c91887...-function=0x401A77-api(kernel32.WriteFile)-True
SKIP kernel32-file-export(BaseThreadInitThunk)-True
SKIP kernel32-file-export(lstrlenW)-True
SKIP kernel32-file-export(nope)-False
SKIP kernel32-64-function=0x180001010-api(RtlVirtualUnwind)-True
SKIP kernel32-64-function=0x180001010-api(RtlVirtualUnwind)-True
SKIP kernel32-64-function=0x180001068-characteristic(gs access)-True
SKIP kernel32-64-function=0x180001068-characteristic(cross section flow)-False
SKIP kernel32-64-function=0x1800017D0-characteristic(peb access)-True
SKIP kernel32-64-function=0x1800202B0-api(RtlCaptureContext)-True
SKIP kernel32-64-function=0x1800202B0-api(RtlCaptureContext)-True
OK   mimikatz-file-string(SCardControl)-True
OK   mimikatz-file-string(SCardTransmit)-True
OK   mimikatz-file-string(ACR  > )-True
OK   mimikatz-file-string(nope)-False
OK   mimikatz-file-section(.text)-True
OK   mimikatz-file-section(.nope)-False
OK   mimikatz-file-import(advapi32.CryptSetHashParam)-True
OK   mimikatz-file-import(CryptSetHashParam)-True
OK   mimikatz-file-import(kernel32.IsWow64Process)-True
OK   mimikatz-file-import(msvcrt.exit)-True
OK   mimikatz-file-import(cabinet.#11)-True
OK   mimikatz-file-import(#11)-False
OK   mimikatz-file-import(#nope)-False
OK   mimikatz-file-import(nope)-False
OK   mimikatz-file-import(advapi32.CryptAcquireContextW)-True
OK   mimikatz-file-import(advapi32.CryptAcquireContext)-True
OK   mimikatz-file-import(CryptAcquireContextW)-True
OK   mimikatz-file-import(CryptAcquireContext)-True
OK   mimikatz-function=0x401000-characteristic(loop)-False
OK   mimikatz-function=0x401000-characteristic(tight loop)-False
OK   mimikatz-function=0x401000-characteristic(stack string)-False
OK   mimikatz-function=0x401000-number(0x0)-True
OK   mimikatz-function=0x401000,bb=0x401000-characteristic(tight loop)-False
OK   mimikatz-function=0x40105D-mnemonic(push)-True
OK   mimikatz-function=0x40105D-mnemonic(movzx)-True
OK   mimikatz-function=0x40105D-mnemonic(xor)-True
OK   mimikatz-function=0x40105D-mnemonic(in)-False
OK   mimikatz-function=0x40105D-mnemonic(out)-False
OK   mimikatz-function=0x40105D-number(0xFF)-True
OK   mimikatz-function=0x40105D-number(0x3136B0)-True
OK   mimikatz-function=0x40105D-number(0xC)-False
OK   mimikatz-function=0x40105D-number(0x10)-False
OK   mimikatz-function=0x40105D-offset(0x0)-True
OK   mimikatz-function=0x40105D-offset(0x4)-True
OK   mimikatz-function=0x40105D-offset(0xC)-True
OK   mimikatz-function=0x40105D-offset(0x8)-False
OK   mimikatz-function=0x40105D-offset(0x10)-False
OK   mimikatz-function=0x40105D-string(SCardControl)-True
OK   mimikatz-function=0x40105D-string(SCardTransmit)-True
OK   mimikatz-function=0x40105D-string(ACR  > )-True
OK   mimikatz-function=0x40105D-string(nope)-False
OK   mimikatz-function=0x40105D-bytes(53 00 43 00 61 00 72 00 64 00 43 00 6F 00 6E 00 74 00 72 00 6F 00 6C 00)-True
OK   mimikatz-function=0x40105D-bytes(53 00 43 00 61 00 72 00 64 00 54 00 72 00 61 00 6E 00 73 00 6D 00 69 00 74 00)-True
OK   mimikatz-function=0x40105D-bytes(41 00 43 00 52 00 20 00 20 00 3E 00 20 00)-True
OK   mimikatz-function=0x40105D-bytes(6E 6F 70 65)-False
OK   mimikatz-function=0x40105D-characteristic(nzxor)-False
OK   mimikatz-function=0x40105D-characteristic(calls to)-True
OK   mimikatz-function=0x40105D,bb=0x401073-operand[1].number(0xFF)-True
OK   mimikatz-function=0x40105D,bb=0x401073-operand[0].number(0xFF)-False
OK   mimikatz-function=0x40105D,bb=0x4010B0-operand[0].offset(0x4)-True
OK   mimikatz-function=0x40105D,bb=0x4010B0-operand[1].offset(0x4)-False
OK   mimikatz-function=0x4011FB-offset(-0x1)-True
OK   mimikatz-function=0x4011FB-offset(-0x2)-True
OK   mimikatz-function=0x401517-characteristic(loop)-True
OK   mimikatz-function=0x401553-number(0xFFFFFFFF)-True
OK   mimikatz-function=0x401873,bb=0x4018B2,insn=0x4018C0-number(0x2)-True
OK   mimikatz-function=0x401CC7,bb=0x401CDE,insn=0x401CF6-offset(0x10)-False
OK   mimikatz-function=0x401D64,bb=0x401D73,insn=0x401D85-offset(0x80000000)-False
OK   mimikatz-function=0x402203,bb=0x402221,insn=0x40223C-offset(0x4)-True
OK   mimikatz-function=0x402EC4-characteristic(tight loop)-True
OK   mimikatz-function=0x402EC4,bb=0x402F8E-characteristic(tight loop)-True
OK   mimikatz-function=0x403BAC-api(advapi32.CryptAcquireContextW)-True
OK   mimikatz-function=0x403BAC-api(advapi32.CryptAcquireContext)-True
OK   mimikatz-function=0x403BAC-api(advapi32.CryptGenKey)-True
OK   mimikatz-function=0x403BAC-api(advapi32.CryptImportKey)-True
OK   mimikatz-function=0x403BAC-api(advapi32.CryptDestroyKey)-True
OK   mimikatz-function=0x403BAC-api(CryptAcquireContextW)-True
OK   mimikatz-function=0x403BAC-api(CryptAcquireContext)-True
OK   mimikatz-function=0x403BAC-api(CryptGenKey)-True
OK   mimikatz-function=0x403BAC-api(CryptImportKey)-True
OK   mimikatz-function=0x403BAC-api(CryptDestroyKey)-True
OK   mimikatz-function=0x403BAC-api(Nope)-False
OK   mimikatz-function=0x403BAC-api(advapi32.Nope)-False
OK   mimikatz-function=0x40640e-characteristic(recursive call)-True
OK   mimikatz-function=0x40B3C6-api(LocalFree)-True
OK   mimikatz-function=0x410DFC-characteristic(nzxor)-True
OK   mimikatz-function=0x4175FF-characteristic(recursive call)-False
OK   mimikatz-function=0x4175FF-characteristic(indirect call)-True
OK   mimikatz-function=0x43e543-number(0xFFFFFFF0)-True
OK   mimikatz-function=0x44570F-bytes(FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF)-False
OK   mimikatz-function=0x44EDEF-string(INPUTEVENT)-True
OK   mimikatz-function=0x44EDEF-bytes(49 00 4E 00 50 00 55 00 54 00 45 00 56 00 45 00 4E 00 54 00)-True
OK   mimikatz-function=0x4556E5-characteristic(stack string)-True
OK   mimikatz-function=0x4556E5-api(advapi32.LsaQueryInformationPolicy)-True
OK   mimikatz-function=0x4556E5-api(LsaQueryInformationPolicy)-True
OK   mimikatz-function=0x4556E5-characteristic(peb access)-False
OK   mimikatz-function=0x4556E5-characteristic(gs access)-False
OK   mimikatz-function=0x4556E5-characteristic(cross section flow)-False
OK   mimikatz-function=0x4556E5-characteristic(indirect call)-False
OK   mimikatz-function=0x4556E5-characteristic(calls from)-True
OK   mimikatz-function=0x456BB9-characteristic(calls to)-False
OK   mimikatz-function=0x46D534-characteristic(nzxor)-False
OK   mimikatz-function=0x46D6CE-string((null))-True
OK   mimikatz-function=0x4702FD-characteristic(calls from)-False
OK   mimikatz-function=0x47153B,bb=0x4717AB,insn=0x4717B1-number(-0x30)-False
OK   mimikatz-function=0x471EAB,bb=0x471ED8,insn=0x471EE6-number(0x4)-False
SKIP pma12-04-file-characteristic(embedded pe)-True
SKIP pma16-01-file-function-name(__aulldiv)-True
SKIP pma16-01-file-os(windows)-True
SKIP pma16-01-file-os(linux)-False
SKIP pma16-01-file-arch(i386)-True
SKIP pma16-01-file-arch(amd64)-False
SKIP pma16-01-file-format(pe)-True
SKIP pma16-01-file-format(elf)-False
SKIP pma16-01-function=0x4021B0-regex(string =~ HTTP/1.0)-True
SKIP pma16-01-function=0x402F40-regex(string =~ www.practicalmalwareanalysis.com)-True
SKIP pma16-01-function=0x402F40-substring(practicalmalwareanalysis.com)-True
SKIP pma16-01-function=0x404356-os(windows)-True
SKIP pma16-01-function=0x404356-arch(i386)-True
SKIP pma16-01-function=0x404356,bb=0x4043B9-os(windows)-True
SKIP pma16-01-function=0x404356,bb=0x4043B9-arch(i386)-True
OK   mimikatz-file-import(cabinet.FCIAddFile)-True
DONE
```

### Checklist

- [x] No CHANGELOG update needed

- [ ] No new tests needed

- [ ] No documentation update needed

### CODE REVIEW

**Comment by mr-tz** [comment]
```
Results (1383.94s):
     483 passed
      32 failed
         - tests/test_dnfile_features.py:15 test_dnfile_features[_1c444-function=0x1F59, bb=0x1F59, insn=0x1F5B-characteristic(unmanaged call)-True]
         - tests/test_dnfile_features.py:15 test_dnfile_features[_1c444-function=0x1F68-api(GetWindowDC)-True]
         - tests/test_dnfile_features.py:15 test_dnfile_features[_1c444-function=0x1F68-api(user32.GetWindowDC)-True]
         - tests/test_dnfile_features.py:15 test_dnfile_features[_1c444-function=0x1F68-number(0xCC0020)-True]
         - tests/test_dnfile_features.py:15 test_dnfile_features[_1c444-function=0x1F68-number(0x0)-True]
         - tests/test_dnfile_features.py:15 test_dnfile_features[_1c444-function=0x1F68-number(0x1)-False]
         - tests/test_dnfile_features.py:15 test_dnfile_features[_1c444-function=0x1F68, bb=0x1F68, insn=0x1FF9-api(System.Drawing.Image::FromHbitmap)-True]
         - tests/test_dnfile_features.py:15 test_dnfile_features[_1c444-function=0x1F68, bb=0x1F68, insn=0x1FF9-api(FromHbitmap)-False]
         - tests/test_dnfile_features.py:15 test_dnfile_features[_1c444-function=0x2544-characteristic(unmanaged call)-False]
         - tests/test_dnfile_features.py:15 test_dnfile_features[hello-world-function=0x250-string(Hello World!)-True]
         - tests/test_dnfile_features.py:15 test_dnfile_features[hello-world-function=0x250-api(System.Console::WriteLine)-True]
         - tests/test_dnfile_features.py:15 test_dnfile_features[hello-world-function=0x250, bb=0x250, insn=0x252-string(Hello World!)-True]
         - tests/test_dotnet_features.py:15 test_dnfile_features[_1c444-function=0x1F59, bb=0x1F59, insn=0x1F5B-characteristic(unmanaged call)-True]
         - tests/test_dotnet_features.py:15 test_dnfile_features[_1c444-function=0x1F68-api(GetWindowDC)-True]
         - tests/test_dotnet_features.py:15 test_dnfile_features[_1c444-function=0x1F68-api(user32.GetWindowDC)-True]
         - tests/test_dotnet_features.py:15 test_dnfile_features[_1c444-function=0x1F68-number(0xCC0020)-True]
         - tests/test_dotnet_features.py:15 test_dnfile_features[_1c444-function=0x1F68-number(0x0)-True]
         - tests/test_dotnet_features.py:15 test_dnfile_features[_1c444-function=0x1F68-number(0x1)-False]
         - tests/test_dotnet_features.py:15 test_dnfile_features[_1c444-function=0x1F68, bb=0x1F68, insn=0x1FF9-api(System.Drawing.Image::FromHbitmap)-True]
         - tests/test_dotnet_features.py:15 test_dnfile_features[_1c444-function=0x1F68, bb=0x1F68, insn=0x1FF9-api(FromHbitmap)-False]
         - tests/test_dotnet_features.py:15 test_dnfile_features[_1c444-function=0x2544-characteristic(unmanaged call)-False]
         - tests/test_dotnet_features.py:15 test_dnfile_features[hello-world-function=0x250-string(Hello World!)-True]
         - tests/test_dotnet_features.py:15 test_dnfile_features[hello-world-function=0x250-api(System.Console::WriteLine)-True]
         - tests/test_dotnet_features.py:15 test_dnfile_features[hello-world-function=0x250, bb=0x250, insn=0x252-string(Hello World!)-True]
         - tests/test_freeze.py:59 test_null_feature_extractor
         - tests/test_freeze.py:171 test_freeze_sample
         - tests/test_freeze.py:178 test_freeze_load_sample
         - tests/test_main.py:22 test_main
         - tests/test_main.py:420 test_backend_option
         - tests/test_main.py:436 test_json_meta
         - tests/test_scripts.py:34 test_scripts[show-features.py-args5]
         - tests/test_scripts.py:52 test_bulk_process
     129 xfailed
       3 skipped

````

**Comment by mr-tz** [comment]
just pushed an update on dnfile support, running tests locally now

File: capa/features/address.py [hunk 1]
```diff
@@ -27,6 +27,9 @@ def __new__(cls, v):
         assert v >= 0
         return int.__new__(cls, v)
 
+    def __str__(self):
+        return str(self)
```
Comment (williballenthin) [inline capa/features/address.py:31]: what does this do? im surprised it doesn't work already?

Comment (mr-tz) [reply capa/features/address.py:31]: mypy was complaining about this in CI: https://github.com/mandiant/capa/runs/6410100333?check_suite_focus=true#step:8:8
```
...
capa/features/extractors/ida/insn.py:446: error: Cannot instantiate abstract class "AbsoluteVirtualAddress" with abstract attribute "__str__"
capa/features/extractors/ida/function.py:23: error: Cannot instantiate abstract class "AbsoluteVirtualAddress" with abstract attribute "__str__"
capa/features/extractors/common.py:24: error: Cannot instantiate abstract class "FileOffsetAddress" with abstract attribute "__str__"
capa/features/extractors/common.py:27: error: Cannot instantiate abstract class "FileOffsetAddress" with abstract attribute "__str__"
capa/features/extractors/viv/file.py:25: error: Cannot instantiate abstract class "FileOffsetAddress" with abstract attribute "__str__"
...
```

**Pull request merged**